From 3a7a7a86b893a5dbbb365800de76bdbce9a795fc Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:20:09 -0400
Subject: [fs] nfsd: Decrease nfsd_users in nfsd_startup_generic fail

Message-id: <1410898849-22126-126-git-send-email-bfields@redhat.com>
Patchwork-id: 95760
O-Subject: [Patch RHEL7 125/165] NFSD: Decrease nfsd_users in nfsd_startup_generic fail
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Kinglong Mee <kinglongmee@gmail.com>

A memory allocation failure could cause nfsd_startup_generic to fail, in
which case nfsd_users wouldn't be incorrectly left elevated.

After nfsd restarts nfsd_startup_generic will then succeed without doing
anything--the first consequence is likely nfs4_start_net finding a bad
laundry_wq and crashing.

Signed-off-by: Kinglong Mee <kinglongmee@gmail.com>
Fixes: 4539f14981ce "nfsd: replace boolean nfsd_up flag by users counter"
Cc: stable@vger.kernel.org
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: d9499a95716db0d4bc9b67e88fd162133e7d6b08
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfssvc.c b/fs/nfsd/nfssvc.c
index 5d026dc..752d56b 100644
--- a/fs/nfsd/nfssvc.c
+++ b/fs/nfsd/nfssvc.c
@@ -221,7 +221,8 @@ static int nfsd_startup_generic(int nrservs)
   */
  ret = nfsd_racache_init(2*nrservs);
  if (ret)
-  return ret;
+  goto dec_users;
+
  ret = nfs4_state_start();
  if (ret)
   goto out_racache;
@@ -229,6 +230,8 @@ static int nfsd_startup_generic(int nrservs)
 
 out_racache:
  nfsd_racache_shutdown();
+dec_users:
+ nfsd_users--;
  return ret;
 }
 
-- 
1.7.1